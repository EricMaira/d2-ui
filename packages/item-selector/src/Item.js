import React from 'react';
import PropTypes from 'prop-types';
import { colors } from './colors';
import RemoveDimensionButton from './buttons/RemoveDimensionButton';
import ItemIcon from './ItemIcon';
import { styles } from './styles/Item.style';

const Icon = ({ selected, highlighted }) => {
    const { grey, white, accentSecondary } = colors;
    const bgColor = !selected ? grey : highlighted ? white : accentSecondary;

    return <ItemIcon backgroundColor={bgColor} />;
};

const onClickWrapper = (id, index, onClick) => event =>
    onClick(event.metaKey || event.ctrlKey, event.shiftKey, index, id);

export const Item = ({
    selected,
    highlighted,
    id,
    index,
    onRemoveItem,
    onClick,
    name,
    isGhost,
}) => {
    const selectedState = selected ? 'selected' : 'unselected';
    // const divClassName = [`${selectedState}-list-item`]
    //     .concat(isGhost ? 'ghost' : '')
    //     .join(' ');

    //
    const highlightedStyle = highlighted ? styles.highlightedItem : {};
    const itemStyle = selected ? Object.assign({}, styles.selectedItem, styles.item) : styles.item;
    const ghostStyle = isGhost ? styles.ghost : {};
    const divStyle = Object.assign({}, itemStyle, ghostStyle, highlightedStyle);

    const highlightedSpanStyle = highlighted ? styles.highlightedText : {};
    const labelStyle = selected ? styles.selectedItemLabel : styles.unselectedItemLabel;
    const spanStyle = Object.assign({}, labelStyle, highlightedSpanStyle);


    return (
        <div
            data-test={`dimension-item-${id}`}
            style={divStyle}
            onClick={onClickWrapper(id, index, onClick)}
        >
            <Icon selected={selected} highlighted={highlighted} />
            <span style={spanStyle}>{name}</span>
            {selected && (
                <RemoveDimensionButton
                    highlighted={highlighted}
                    onClick={() => onRemoveItem(id)}
                />
            )}
        </div>
    );
};

Item.defualtProps = {
    onRemoveItem: () => null,
    onClick: () => null,
};

Item.propTypes = {
    id: PropTypes.string.isRequired,
    index: PropTypes.number.isRequired,
    name: PropTypes.string.isRequired,
    highlighted: PropTypes.bool.isRequired,
    onClick: PropTypes.func,
    onRemoveItem: PropTypes.func,
    selected: PropTypes.bool,
    isGhost: PropTypes.bool,
};

export default Item;
