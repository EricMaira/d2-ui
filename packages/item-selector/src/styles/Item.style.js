import { colors } from '../colors';

export const styles = {
    highlightedItem: {
        backgroundColor: colors.accentSecondaryDark,
    },
    highlightedText: {
        color: colors.white,
    },
    item: {
        display: 'flex',
        alignItems: 'center',
        padding: '3px',
        borderRadius: '4px',
        cursor: 'pointer',
    },
    ghost: {
        transition: 'opacity 0.2s ease',
        opacity: '0.2',
    },
    selectedItem: {
        backgroundColor: '#b2dfdb',
    },
    selectedItemLabel: {
        fontSize: '14px',
        padding: '2px',
    },
    unselectedItemLabel: {
        fontSize: '14px',
        padding: '2px 5px 2px 2px',
    }
};
