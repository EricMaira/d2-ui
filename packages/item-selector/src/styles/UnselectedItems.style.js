export const styles = {
    selectButton: {
        margin: '0 auto',
        display: 'block',
    },
    unselectedItems: { flex: '1', overflowY: 'auto' },
    unselectedItemsList: {
        userSelect: 'none',
        listStyle: 'none',
        borderBottom: '0px',
        paddingLeft: '0px',
        margin: '0px',
    },
    listItem: {
        display: 'flex',
        margin: '2px',
    },
    assignButton: {
        position: 'absolute',
        left: '429px',
        top: '230px',
    }
};
