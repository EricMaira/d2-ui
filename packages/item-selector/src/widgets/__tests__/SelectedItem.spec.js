import React from 'react';
import { shallow } from 'enzyme';
import Item from '../SelectedItem';
import DeselectIconButton from '../DeselectIconButton';
import ItemIcon from '../ItemIcon';
import { colors } from '../../styles/colors';

describe('The Item component ', () => {
    let props;
    let shallowItem;

    const item = () => {
        if (!shallowItem) {
            shallowItem = shallow(<Item {...props} />);
        }
        return shallowItem;
    };

    beforeEach(() => {
        props = {
            id: 'testID',
            name: 'The selected item',
            index: 0,
            highlighted: false,
            ghost: false,
            onClick: jest.fn(),
            onRemove: jest.fn(),
        };
        shallowItem = undefined;
    });

    it('Should match the snapshot', () => {
        const wrapper = item();

        expect(wrapper).toMatchSnapshot();
    });

    // it('renders a div containing everything else', () => {
    //     const wrappingDiv = item()
    //         .find('div')
    //         .first();

    //     expect(
    //         item()
    //             .find('div')
    //             .first()
    //     ).toHaveLength(1);
    //     expect(wrappingDiv.children()).toEqual(item().children());
    // });

    // it('renders <UnselectedIcon /> when prop "selected" is true ', () => {
    //     const icon = item()
    //         .find('Icon')
    //         .dive()
    //         .find(ItemIcon);

    //     expect(icon.prop('backgroundColor')).toEqual(colors.grey);
    // });

    // it('renders a <HighlightedIcon /> when props highlighted and selected are true', () => {
    //     props.highlighted = true;
    //     props.selected = true;

    //     const icon = item()
    //         .find('Icon')
    //         .dive()
    //         .find(ItemIcon);

    //     expect(icon.prop('backgroundColor')).toEqual(colors.white);
    // });

    // it('renders <SelectedIcon /> when className is equal to "selected" ', () => {
    //     props.selected = true;

    //     const icon = item()
    //         .find('Icon')
    //         .dive()
    //         .find(ItemIcon);

    //     expect(icon.prop('backgroundColor')).toEqual(colors.accentSecondary);
    // });

    // it('renders <DeselectIconButton /> for selected item ', () => {
    //     props.selected = true;

    //     const removeButton = item().find(DeselectIconButton);

    //     expect(removeButton.length).toEqual(1);
    // });

    describe('onClick', () => {
        it('fires onClick property', () => {
            item()
                .props()
                .onClick({ preventDefault: () => undefined });

            expect(props.onClick).toBeCalledTimes(1);
        });

        it('fires onClick with correct arguments when metaKey pressed', () => {
            item()
                .props()
                .onClick({
                    preventDefault: () => undefined,
                    metaKey: true,
                    ctrlKey: false,
                    shiftKey: false,
                });

            expect(props.onClick).toBeCalledTimes(1);
            expect(props.onClick).toBeCalledWith(true, false, 0, 'testID');
        });

        it('fires onClick with correct arguments when ctrlKey pressed', () => {
            item()
                .props()
                .onClick({
                    preventDefault: () => undefined,
                    metaKey: false,
                    ctrlKey: true,
                    shiftKey: false,
                });

            expect(props.onClick).toBeCalledTimes(1);
            expect(props.onClick).toBeCalledWith(true, false, 0, 'testID');
        });

        it('fires onClick with correct arguments when shiftKey pressed', () => {
            item()
                .props()
                .onClick({
                    preventDefault: () => undefined,
                    metaKey: false,
                    ctrlKey: false,
                    shiftKey: true,
                });

            expect(props.onClick).toBeCalledTimes(1);
            expect(props.onClick).toBeCalledWith(false, true, 0, 'testID');
        });
    });
});
