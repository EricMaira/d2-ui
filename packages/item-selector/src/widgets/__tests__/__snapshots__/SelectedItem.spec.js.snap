// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Item component  Should match the snapshot 1`] = `
<div
  className="jsx-3290541332 item selected-item"
  data-test="dimension-item-testID"
  onClick={[Function]}
>
  <ItemIcon
    backgroundColor="#00796B"
  />
  <span
    className="jsx-3290541332 item-label"
  >
    The selected item
  </span>
  <DeselectIconButton
    fill="#00796B"
    onClick={[Function]}
  />
  <JSXStyle
    id="3290541332"
  >
    .
    h
    i
    g
    h
    l
    i
    g
    h
    t
    e
    d
    -
    t
    e
    x
    t
    .
    j
    s
    x
    -
    3
    2
    9
    0
    5
    4
    1
    3
    3
    2
    {
    c
    o
    l
    o
    r
    :
    #
    F
    F
    F
    F
    F
    F
    ;
    }
    .
    i
    c
    o
    n
    .
    j
    s
    x
    -
    3
    2
    9
    0
    5
    4
    1
    3
    3
    2
    {
    b
    a
    c
    k
    g
    r
    o
    u
    n
    d
    -
    c
    o
    l
    o
    r
    :
    #
    0
    0
    7
    9
    6
    B
    ;
    }
    .
    i
    t
    e
    m
    .
    j
    s
    x
    -
    3
    2
    9
    0
    5
    4
    1
    3
    3
    2
    {
    -
    w
    e
    b
    k
    i
    t
    -
    a
    l
    i
    g
    n
    -
    i
    t
    e
    m
    s
    :
    c
    e
    n
    t
    e
    r
    ;
    -
    w
    e
    b
    k
    i
    t
    -
    b
    o
    x
    -
    a
    l
    i
    g
    n
    :
    c
    e
    n
    t
    e
    r
    ;
    -
    m
    s
    -
    f
    l
    e
    x
    -
    a
    l
    i
    g
    n
    :
    c
    e
    n
    t
    e
    r
    ;
    a
    l
    i
    g
    n
    -
    i
    t
    e
    m
    s
    :
    c
    e
    n
    t
    e
    r
    ;
    b
    o
    r
    d
    e
    r
    -
    r
    a
    d
    i
    u
    s
    :
    4
    p
    x
    ;
    c
    u
    r
    s
    o
    r
    :
    p
    o
    i
    n
    t
    e
    r
    ;
    d
    i
    s
    p
    l
    a
    y
    :
    -
    w
    e
    b
    k
    i
    t
    -
    b
    o
    x
    ;
    d
    i
    s
    p
    l
    a
    y
    :
    -
    w
    e
    b
    k
    i
    t
    -
    f
    l
    e
    x
    ;
    d
    i
    s
    p
    l
    a
    y
    :
    -
    m
    s
    -
    f
    l
    e
    x
    b
    o
    x
    ;
    d
    i
    s
    p
    l
    a
    y
    :
    f
    l
    e
    x
    ;
    p
    a
    d
    d
    i
    n
    g
    :
    3
    p
    x
    ;
    }
    .
    g
    h
    o
    s
    t
    .
    j
    s
    x
    -
    3
    2
    9
    0
    5
    4
    1
    3
    3
    2
    {
    o
    p
    a
    c
    i
    t
    y
    :
    0
    .
    2
    ;
    -
    w
    e
    b
    k
    i
    t
    -
    t
    r
    a
    n
    s
    i
    t
    i
    o
    n
    :
    o
    p
    a
    c
    i
    t
    y
     
    0
    .
    2
    s
     
    e
    a
    s
    e
    ;
    t
    r
    a
    n
    s
    i
    t
    i
    o
    n
    :
    o
    p
    a
    c
    i
    t
    y
     
    0
    .
    2
    s
     
    e
    a
    s
    e
    ;
    }
    .
    s
    e
    l
    e
    c
    t
    e
    d
    -
    i
    t
    e
    m
    .
    j
    s
    x
    -
    3
    2
    9
    0
    5
    4
    1
    3
    3
    2
    {
    b
    a
    c
    k
    g
    r
    o
    u
    n
    d
    -
    c
    o
    l
    o
    r
    :
    #
    b
    2
    d
    f
    d
    b
    ;
    }
    .
    s
    e
    l
    e
    c
    t
    e
    d
    -
    i
    t
    e
    m
    .
    j
    s
    x
    -
    3
    2
    9
    0
    5
    4
    1
    3
    3
    2
    :
    h
    o
    v
    e
    r
    {
    b
    a
    c
    k
    g
    r
    o
    u
    n
    d
    -
    c
    o
    l
    o
    r
    :
    #
    8
    6
    c
    4
    b
    f
    ;
    }
    .
    h
    i
    g
    h
    l
    i
    g
    h
    t
    e
    d
    -
    i
    t
    e
    m
    .
    j
    s
    x
    -
    3
    2
    9
    0
    5
    4
    1
    3
    3
    2
    {
    b
    a
    c
    k
    g
    r
    o
    u
    n
    d
    -
    c
    o
    l
    o
    r
    :
    #
    0
    0
    4
    C
    4
    0
    ;
    }
    .
    i
    t
    e
    m
    -
    l
    a
    b
    e
    l
    .
    j
    s
    x
    -
    3
    2
    9
    0
    5
    4
    1
    3
    3
    2
    {
    f
    o
    n
    t
    -
    s
    i
    z
    e
    :
    1
    4
    p
    x
    ;
    p
    a
    d
    d
    i
    n
    g
    :
    2
    p
    x
    ;
    }
  </JSXStyle>
</div>
`;
